### DISCUSSION TITLE
How is [hash:base64] generated when using CSS modules and css-loader?

### DISCUSSION THREAD

**Discussion by JoshuaCrewe** [webpack/webpack#17350] [Q&A] [answered]
I have tried to find out the answer to this on my own and haven't had the skills to get a reliable result.

We are using Laravel Mix to compile our assets and there appears to be a bug which doesn't allow defining `localIdentName`. This means that it _should_ be using the default `[hash:base64]` .

The question is, how is that hash actually generated? I thought it was doing a base64 hash of the class name with perhaps a salt but I couldn't find the salt and I couldn't recreate the generated class names by base64 encoding a variety of source strings. I tried :
- The class name
- class name + path
- class name + src
- Searching through code to find the string which is hashed
- Looking for where the base64 encode happens

My hope is that if I know how the hashes are created I can recreate the process in PHP and enable scoped styles that way. Also this has now become a scratch to itch and learn how this is done. Any input/help is greatly appreciated.

https://github.com/laravel-mix/laravel-mix/issues/3359

**Comment by JoshuaCrewe** [comment]
Picking the thread apart a bit more I think the actual hashing is done in https://github.com/webpack/loader-utils/blob/0c98405f9ab36b26b2f47843c50023199c440a6d/lib/getHashDigest.js#L63 .

The buffer tells me what is the source string, Then it is hashed (which I don't understand) and then base64 encoded.

By following the logic this should produce the same result :

```php
$hash = hash('md4', 'src/app.module.cssslideshow', true);
$base64 = base64_encode($hash);
```
But it doesn't which means my logic is faulty :smile:

**Reply by alexander-akait** [reply]
It is not simple to recreate it (but possible), why you need to do this?

**Reply by JoshuaCrewe** [reply]
My first desire was to change the template and drop the hash, something like `namespace_[src]_[local]`. But there is a bug in laravel mix (I think) which isn't allowing me to do this.

Next thought was if I could recrete what is being produced by default then I could work with that. Use CSS modules to change the class names and then in PHP/Twig use a function to do the same conversion plus a key to keep it human readable.

Reading that it was base64 encoded made me think it would be possible and not that hard, so I appreciate your input. 

If it is a waste of time I should either try and fix the bug, switch to postcss and use a plugin or start my own webpack config. Ironically I didn't do some of these things to save time :)

Fixing the template in laravel mix would be ideal but I need help as I can't figure it out on my own

**Reply by alexander-akait** [reply]
Based on it https://github.com/laravel-mix/laravel-mix/issues/3359, yes, laravel-mix has a bug and don't allow to setup some options, fix will be ideal here, because recreate logic is not easy, you need to get filenamd and context and classes and them run the same logic for generation

**Reply by JoshuaCrewe** [reply]
Thanks for weighing in, I appreciate it. I my focus would be on the bug fix for sure.

I was looking at the logic which is used to generate the scoped classes and it didn't look too bad. Once you have the string (made up of filename, path and class) it appears to get an md4 hash to hash the string and base64 encode the outcome of that.

I am new to the node crypto library and I couldn't get it to print out the hash that is generated to test against the PHP version I made. Is the crypto md4 hash hard to reproduce because the hash is generated based on the environment?

**Reply by JoshuaCrewe** [reply]
It is helpful to type out how I understand it so that I can see faulty logic or spark the learning process.

In `loader-utils` there is the function `getHashDigest` which is where the magic happens. We can follow the logic so far : 
```
  console.log(`the buffer is ${buffer} `); // src/file/class
  console.log(`the HashType is ${hashType} `); // md4
  console.log(`the digestType is ${digestType} `); // base64
  console.log(`the maxLength is ${maxLength} `); // 9999
```
But then it gets to the hashing part : 
```
hash = require("crypto").createHash(hashType);
...
hash.update(buffer);
...
return hash.digest(digestType || "hex").substr(0, maxLength);
``` 

I am unfamiliar with the hashing object which requires me reading the documentation closer, but the above three commands result in the scoped class name. It appears to : 
1. Create a hash using md4
2. use that hash object to transform the string
3. base64 encode this to get the scoped class name

I thought the equivalent in PHP would have been : 
```
$hash = hash('md4', 'src/app.module.cssslideshow', true);
$base64 = base64_encode($hash);
```
But the end result is different. The only variable is the hash that node generates, but if this can be replicated in PHP it should be OK. The question becomes how to translate the md4 hash in PHP then.

![2023-06-09_15-15](https://github.com/webpack/webpack/assets/12238901/cfa7e5b3-269c-4b55-ba12-0eedb8988168)

**Reply by alexander-akait** [reply]
There is a full logic https://github.com/webpack-contrib/css-loader/blob/master/src/utils.js#L327, also you miss this one https://github.com/webpack-contrib/css-loader/blob/master/src/utils.js#L351, we use unprintable character between local and module

**Reply by JoshuaCrewe** [reply]
Ooo the unprintable character could be the thing which is trowing it off. I will try and find a way to make sure that is included in the process 🤞

**Reply by JoshuaCrewe** [reply]
Yep. That is the missing piece of this puzzle. There is the unprintable character which I missed when doing my rudimentary debugging. This means that the difference I was getting in my results were because the source string being hashed was different.

**Comment by JoshuaCrewe** [comment, marked as answer]
Thank you so much to @alexander-akait , I have been able to follow the logic and recreate the way css-loader generates a class name hash for CSS modules by default.

1. The string which is being hashed consists of the path, file name, unprintable character and the class name
2. This string is hashed using md4
3. This hash is base64 encoded
4. If the resulting string starts with a number, an underscore is prepended
5. Forward slashes are replaced with dashes.

Early POC this PHP appears to recreate the generated class names : 
```
class ClassName
{
    private $path = 'public/content/themes/brickblocks/modules/components/card/';
    private $file = 'style.module.scss';
    private $sep = "\x00";

    public function generateClass($class)
    {
        // Build the string to hash
        $string =  "{$this->path}{$this->file}{$this->sep}{$class}";

        // Generate a hash
        $hash = hash('md4', $string, true);

        // Encode it for easier use on the web
        $base64 = base64_encode($hash);

        // HTML class names cannot start with a number
        $fstchar = substr($base64, 0, 1);
        if (is_numeric($fstchar)) {
            $base64 = '_' . $base64; // Add an underscore
        }

        // Forward slashes get replaces with a dash.
        $base64 = str_replace('/', '-', $base64);

        return $base64;
    }
}
```
